storage/src/test/java/org/apache/kafka/storage/internals/log/LogConcurrencyTest.java

```
@@ -0,0 +1,237 @@
+            final LogAppendTask appendTask = new LogAppendTask(log, maxOffset);
+
+            final Future<?> consumerFuture = executor.submit(consumer);
+            final Future<?> fetcherTaskFuture = executor.submit(appendTask);
+
+            fetcherTaskFuture.get();
+            consumerFuture.get();
+
+            validateConsumedData(log, consumer.getConsumedBatches());
+        } finally {
+            executor.shutdownNow();
+        }
+    }
+
+    private UnifiedLog createLog() throws IOException {
+        return createLog(new LogConfig(new Properties()));
+    }
+
+    private UnifiedLog createLog(LogConfig config) throws IOException {
+        log = UnifiedLog.create(
+            logDir,
+            config,
+            0L,
+            0L,
+            scheduler,
+            brokerTopicStats,
+            Time.SYSTEM,
+            5 * 60 * 1000,
+            new ProducerStateManagerConfig(TransactionLogConfig.PRODUCER_ID_EXPIRATION_MS_DEFAULT, false),
+            TransactionLogConfig.PRODUCER_ID_EXPIRATION_CHECK_INTERVAL_MS_DEFAULT,
+            new LogDirFailureChannel(10),
+            true,
+            Optional.empty()
+        );
+
+        return log;
+    }
+
+    private void validateConsumedData(UnifiedLog log, List<FetchedBatch> consumedBatches) {
+        final Iterator<FetchedBatch> iter = consumedBatches.iterator();
+        log.logSegments().forEach(segment ->
+            segment.log().batches().forEach(batch -> {
+                if (iter.hasNext()) {
+                    final FetchedBatch consumedBatch = iter.next();
+                    try {
+                        assertEquals(batch.partitionLeaderEpoch(),
+                            consumedBatch.epoch(), "Consumed batch with unexpected leader epoch");
+                        assertEquals(batch.baseOffset(),
+                            consumedBatch.baseOffset(), "Consumed batch with unexpected base offset");
+                    } catch (Throwable t) {
```

we could throw the exception directly

---

Great.